[WIP] File set combinators

infinisil opened the pull request:
This PR rethinks @roberth's original [source combinators](https://github.com/NixOS/nixpkgs/pull/112083) into a more composable datatype: Sets of files. Nothing is decided yet, but I'm really amazed by how well this is working.

The abstraction here allows efficient and lazy operations on a set of files. Typical set operations are supported, including union, intersection, difference and filtering.

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles: 

Update: This has now been mostly incrementally merged! See https://github.com/NixOS/nixpkgs/issues/266356 for more info.

### Plan
1. [x] Get the code fully working, but without fully commented code
2. [x] Write clean introduction and reference documentation
   1. [x] Reference
   2. [x] Introduction
4. [x] Make a Discourse post, showing off the concept and how people can try it out, to get wider feedback: https://discourse.nixos.org/t/easy-source-filtering-with-file-sets/29117
6. [ ] ~~Iterate the above until everybody is happy with the design~~
    Despite reaching out to the community, not much feedback was given. This does mean that there's no complaints about it either though. Because of this, I'll move towards implementing it, see https://github.com/NixOS/nixpkgs/pull/222981#issuecomment-1641110172
8. [x] Split up this PR into smaller more incremental PR's, see https://github.com/NixOS/nixpkgs/issues/266356

infinisil commented on doc/functions/fileset.section.md:
````diff
@@ -0,0 +1,117 @@
+A fileset can be used in a derivation build by passing it as a `src` to any `stdenv.mkDerivation`-based builder. It is then necessary to also set `sourceRoot` to the path of the builder entry point.
+```nix
+stdenv.mkDerivation {
+  src =
+    with lib.fileset;
+    unions [
+      ./Makefile
+      (suffixFileFilter ["c" "h"] ./src)
+    ];
+  sourceRoot = ./.;
````
Re-using `src` and `sourceRoot` is one of the key design decisions here. It makes a lot of sense and feels decently natural to people that already know about these attributes. `sourceRoot` is required here. I'd still need to update reference docs for this.

Some alternatives:
```nix
mkDerivation {
  src = with lib.fileset;
    toSource ./. (union [
      ./Makefile
      ./src
    ];
}
```

Problems:
- Not very clear that `./.` is just the entry point
- It would have to change the semantics of what it means to assign a string(-coercible) value to `src`. Previously it would always just copy the directory that string points to and make it writable. But with this it can copy not just the directory but also some parents of it, which is somewhat unintuitive. Not sure if that's a big problem though.

```nix
mkDerivation {
  entryPoint = ./.;
  fileset = with lib.fileset;
    union [
      ./Makefile
      ./src
    ];
}
```

Not a big fan of this, `src` is a clear indicator that we're specifying the source, whereas it's not as clear here.

```nix
mkDerivation {
  src.entryPoint = ./.;
  src.fileset = with lib.fileset;
    union [
      ./Makefile
      ./src
    ];
}
```

Not bad.

infinisil commented on doc/functions/fileset.section.md:
````diff
@@ -0,0 +1,117 @@
+}
+```
+
+Warning: Any intermediate directory names will influence the store hash.
````
This is very related to the problem of having to specify the source name as `source` to avoid the repository base name from influencing the store hash. Here it's somewhat unavoidable though. I don't think this is a big problem though, because the user effectively asked for it.

infinisil commented on lib/fileset/default.nix:
```diff
@@ -0,0 +1,175 @@
+  trace = null;
+
+  /*
+  Create a file set containing all files tracked by a local git repository
+
+  Type:
+    gitTracked :: Path -> FileSet
```
Essentially  a `fetchGit ./.`, but returning a file set, useful to `intersect` with to only include git-tracked files.

An alternative would be having users write

```nix
let
  # Necessary for fast evaluation
  predicate = fileIsGitTracked ./.;
in lib.fileset.fileFilter predicate ./src
```

But a `gitTracked` is simpler to use:

```nix
intersect (gitTracked ./.) ./src
```

infinisil commented on lib/fileset/default.nix:
```diff
@@ -0,0 +1,175 @@
+  Limit a file set to only the files satisfying a predicate function.
+
+  Type:
+    fileFilter :: ({
+      path :: Path,
+      root :: Path,
+      components :: [String],
+      name :: String,
+      type :: String,
+      extension :: Null || String,
+      ...
+    } -> Bool) -> FileSet -> FileSet
```
Which attributes should be exposed here? I don't think there's any harm to having all of these, it makes writing file filters easier, though I don't think they will be necessary often.

infinisil commented on doc/functions/fileset.section.md:
```diff
@@ -0,0 +1,117 @@
+      (gitTracked ../.)
+
+      # Remove all files in "manual" directories
+      (lib.flip difference (directoryFilter "manual" ./.)
```
Should we have a `difference` with a flipped order?  The thing to subtract is often smaller than the thing you want to subtract from (as here), so that would be convenient. But what should the name then be?

infinisil commented on lib/fileset/default.nix:
```diff
@@ -0,0 +1,175 @@
+  */
+  difference = null;
+  /*
+  The file set containing all files that are in the first given file set but not in the second.
+
+      differences a [b c] == a ∖ (b ⋃ c)
+
+  Type:
+    differences :: FileSet -> [FileSet] -> FileSet
```
Should this exist? Also relates to the above

infinisil reviewed (commented):
I already got some quick preliminary (generally positive) feedback by @roberth in PMs and meetings.

To move this forward I decided to start writing documentation, even if it's not fully implemented, to get a sense of what the interface should look like. I'm creating comments to highlight some key parts of that interface, but it would be great to get some rough feedback on the other parts of the interface documentation too.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-47/27387/1

infinisil commented:
I've now done a major rewrite of this and am super hype now, this is turning into something really cool. Weekend time now though, I'll continue next week.

This is how I'll try to proceed with this:
1. Get the code fully working, but without perfect error handling and code comments
2. Write clean documentation, at least a reference and a tutorial
3. Make a Discourse post, showing off the concept and how people can try it out, to get wider feedback.
5. Iterate the above until everybody is happy with the design
6. Split up this PR into smaller more incremental PR's, for each:
   1. Write extensive tests
   2. Polish the code
   3. Polish the documentation

roberth commented on doc/functions/fileset.section.md:
````diff
@@ -0,0 +1,130 @@
+}
+```
+
+To see everything you can do with file sets, check out the [reference documentation](#sec-functions-library-fileset).
````
```suggestion
To see everything you can do with file sets, see the [library documentation](#sec-functions-library-fileset).
```

We're already in reference documentation; just an introductory section.

infinisil replied:
I wrote this in a tutorial-style, with the expectation that it would be moved to (or just rendered at) nix.dev at some point (that's where the doc team is moving tutorial docs towards).

roberth replied:
While all text can be learned from, this is not a "learning experience", certainly not according to [diataxis](https://diataxis.fr/tutorials/). And I think that's fine. File sets aren't just functions; the type needs reference documentation too, and it's a good place for a basic introduction to the concept.

> In general, tutorials are the weakest part of documentation, the most misunderstood and the most difficult to do well. Most software projects have poor - or non-existent - tutorials.

infinisil replied:
Well then I clearly don't know how to write a tutorial :sweat_smile:, please take another look at the document now though, I wouldn't really count that as part of the reference.

roberth commented on lib/fileset/default.nix:
```diff
@@ -0,0 +1,580 @@
+    # Strict
+    # TODO: Still needed?
+    # FIXME: This makes equality comparison using `==` much more expensive, which would otherwise work without problems
+    # No, `==` won't ever work well, because we have the recursive directory cutoff thing, so let's not worry about that, `==` is not supported, maybe we should set `_equalityAlwaysReturnsFalse = x: x` because functions always compare false
```
Perhaps better would be `_noEq = throw "filesets are not comparable";`, although that does defeat `deepSeq`. I'd prefer a dialect without broken equality over one with `deepSeq`.

infinisil replied:
Fortunately the `_sparseTree` isn't needed after all, but the `==` problem is still there indeed. I now changed this to a `_noEval = throw "..."`, where the error message gives some alternatives for when equality or printing of file sets is needed

roberth commented on lib/fileset/default.nix:
```diff
@@ -0,0 +1,580 @@
+        Type:
+          gitTracked :: Path -> FileSet
+        */
+        gitTracked = path:
```
This module is about the fileset data type and its operations. This code isn't about that at all; it is rather a consumer of the fileset module, just like git itself is a consumer of the OS file system.
A separate `git` module would be more appropriate, or you could add it later. It is somewhat useful for validating the fileset concept and implementation though.

infinisil replied:
Sounds good, I wasn't sure if it could be implemented without having access to internals, but I figured out it's actually pretty easy using a file filter (currently called `matching`) with a predicate that checks whether the file is in the imported store path.

infinisil replied:
There's a clear separation between core functions and extra convenience functions. The current state of this PR is that _only_ core functions are in `lib.fileset`. However, I do think `lib.fileset` would also be the right place for convenience functions.

In any case `gitTracked` is removed entirely for now. It can't be implemented efficiently without something like https://github.com/NixOS/nix/pull/6530 or https://github.com/NixOS/nix/issues/2944

roberth commented on lib/fileset/default.nix:
```diff
@@ -0,0 +1,580 @@
+          in create (filterTree base "directory");
+
+        /*
+        Limit a file set to only the files for which any path component satisfies some predicate.
```
Matching any component seems a bit odd. Do you have a motivating use case for this?
Also since this is about components, I have a feel for what the last component means, but what determines the first?

infinisil replied:
This again changed a bit: The type of `directoryMatches` is now `(String -> Bool) -> Path -> FileSet`. It doesn't consume a fileset anymore, so this isn't really a combinator anymore, but rather it creates a new file set. The components are now relative to the given `Path`. It still won't match filenames though, because I figured it's easier to combine the result of this with another one for filenames:

```nix
union
  (directoryMatches (name: name == ".git") ./.)
  (filter (file: file.name == ".git") ./.)
```

Rather than the other way around if you need _only_ directories.

With this design, it's clear that `lib.fileset.filter` is only running on files (though maybe it should be `lib.fileset.fileFilter` again, name bikeshedding not yet done), while `lib.fileset.directoryMatches` is only running on directories. This will make it much easier to reason about than the previous `builtins.filterSource`-type functions.

roberth commented on lib/fileset/default.nix:
```diff
@@ -0,0 +1,580 @@
+        /*
+        Whether a path is contained in a file set. If the path points to a directory, the result is whether any file from that directory (recursively) is contained in the file set.
+
+        Same as `intersect path fileset == empty`, but more efficient
```
Not `!= empty`? That's surprising.

infinisil replied:
Ah right, typo in the docs here!

roberth commented on lib/fileset/default.nix:
```diff
@@ -0,0 +1,580 @@
+  # # A directory
+  # <directory> =
+  #   # The inclusion state for every directory entry
+  #   { <name> = <entry>; }
```
When this constructor is passed a single item, it does not add a file to the fileset, so it allows multiple representations of the same set of files.
Some ideas; either
 - _always_ use a smart constructor to prevent this from happening (including operations such as intersection), or
 - change the definition of this data structure to be more than a set, or
 - make sure that the difference is never observable, burdening all the functions that read from this, assuming users don't read the structure themselves

In the latter case, you'd want to return an object with methods only, at least to users, which is a bit inefficient, but perhaps quite nice to use. It also adds a bunch more code, so I don't know whether it'd be a good idea to implement that at this point.

infinisil committed:
lib.lists.commonPrefix[Length]: init

infinisil committed:
Add some functions to lib.path

infinisil committed:
lib.fileset: init

infinisil committed:
lib.fileset: reference introduction docs

infinisil committed:
Integrate sources with subpaths into mkDerivation

Make root, subpath internal

Adapted from https://github.com/NixOS/nixpkgs/pull/112083

Co-Authored-By: Robert Hensing <robert@roberthensing.nl>

infinisil committed:
Slight branch optimization

Avoids a more costly string comparison, favoring an isString instead

infinisil committed:
Improve some error messages and add some code comments

infinisil committed:
Use lib.filesystem.pathType instead lib.sources.pathType

infinisil committed:
Add lib.fileset.optional

infinisil committed:
Minor formatting improvements

infinisil committed:
Improve integration with lib.sources

infinisil committed:
filter -> fileFilter, directoryMatches -> directoryFilter

infinisil committed:
Fix mkDerivation integration

infinisil committed:
Improve importToStore error message

infinisil commented on lib/fileset.nix:
```diff
@@ -0,0 +1,856 @@
+  - `fileset`: The set of files to import into the store.
+    Use the `lib.fileset` combinator functions to define this value.
+
+  - `entryPoint`: The local directory that the resulting directory should focus on, meaning that if you `cd` into the resulting directory, all files in the fileset are accessible at paths relative to the `entryPoint`.
```
Yeah this is hard to explain properly. I should reword this slightly, but the second part of this sentence is not too bad imo:

> meaning that if you `cd` into the resulting directory, all files in the fileset are accessible at paths relative to the `entryPoint`.

An example:

With a file set like this:
```
/home/user/my/project
- frontend
  - index.html
- common
  - vars.h
```

And `entryPoint` being `/home/user/my/project`, this works:
```
cd ${importToStore { ... }}
cat ./index.html ../common/vars.h
```

Underneath, a store path like this is created:
```
/nix/store/...
- frontend
  - index.html
- common
  - vars.h
```

However instead of pointing at it directly, the resulting string (via `outPath`) points at `/nix/store/.../frontend` instead, in order for the `cd` and `cat` to work correctly.

infinisil commented on doc/functions/fileset.section.md:
````diff
@@ -0,0 +1,206 @@
+error: lib.fileset.trace: Expected second argument "/home/user/my/project/non-existent" to be a path that exists, but it doesn't.
+```
+
+File sets can be composed using the functions [`union`](#function-library-lib.fileset.union) (and the list-based equivalent [`unions`](#function-library-lib.fileset.unions)), [`intersect`](#function-library-lib.fileset.intersect) (and the list-based equivalent [`intersects`](#function-library-lib.fileset.intersects)) and [`difference`](#function-library-lib.fileset.difference), the most useful of ...
````
I adapted this from Haskell's `Data.Set` functions, which also features [`union`](https://hackage.haskell.org/package/containers-0.6.7/docs/Data-Set.html#v:union) and [`unions`](https://hackage.haskell.org/package/containers-0.6.7/docs/Data-Set.html#v:union).

We could also read `unions [ a b c ]` as `union a (union b c)`, which then does have multiple `union`s, though that's a bit of a stretch.

Do you perhaps have any suggestions for better names? `unions` is probably one of the most useful functions, so I like how it's currently fairly short.

alyssais replied:
> `unions` is probably one of the most useful functions

Agreed on this, but is `union` (no s) useful?  Maybe we just need the list versions, and then we could call them `union`, `intersect`, etc? In Nix we're a lot less likely to use them as arguments to higher-order functions like `fold` than might be the case in Haskell, where functions that operate on exactly two arguments might be more useful.

infinisil commented on doc/functions/fileset.section.md:
````diff
@@ -0,0 +1,206 @@
+  - main.c (regular)
+```
+
+File sets can be imported into the store using the [`importToStore`](#function-library-lib.fileset.importToStore) function. This function requires the `entryPoint` argument to indicate where the resulting string (via `outPath`, see [`toString`](https://nixos.org/manual/nix/stable/language/builtins.html?highlight=outPath#builtins-toString)) should be focused on, meaning that when you `cd ${importT ...
````
I like it!

infinisil committed:
importToStore -> addToStore

infinisil committed:
Make the mkDerivation src behavior more lazy

roberth commented:
Nix / `lib` sucks at trees.
 - https://github.com/NixOS/nixpkgs/pull/221608#issuecomment-1575458189

It's just a thought; maybe this goes nowhere, but if we could do better at trees, perhaps we could have better file sets, or at least a better file set implementation internally.

infinisil committed:
Add myself as a code owner

infinisil committed:
stdenv.mkDerivation: Implement `srcWorkDir` attribute

Instead of the odd custom file set handling

Entirely independent of file set combinators, only depends on the
existing `lib.sources`

infinisil committed:
lib.fileset.getInfluenceBase: init

infinisil committed:
lib.fileset.traceVal: init

infinisil committed:
lib.path.commonAncestor: init

infinisil committed:
lib.fileset.{impureFromSource -> fromSource}

infinisil committed:
Rethink store importing, lib.fileset.toSource

infinisil committed:
stdenv.mkDerivation: Add fileset integration with `srcFileset`

infinisil committed:
Update reference documentation after recent commits

infinisil committed:
lib.filesystem.pathType: Fix for <2.14, pure eval, nix store paths

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/zurich-23-05-zhf-hackathon-and-workshop-report/29093/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/easy-source-filtering-with-file-sets/29117/1

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+    else
+      readDir path;
+
+  # The following tables are a bit complicated, but they nicely explain the
+  # corresponding implementations, here's the legend:
```
Yes.

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+        lhs' = _directoryEntries path lhs;
+        rhs' = _directoryEntries path rhs;
+      in
+      all (name:
```
Does this take into account different file types?

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+      intermediate;
+
+  # Coerce and normalise the bases of multiple file set values passed to user-facing functions
+  # Type: String -> [ { context :: String, value :: Any } ] -> { commonBase :: Path, trees :: [ <tree> ] }
```
Same here, I think there is some clarity to be gained by shifting away the noise from the non-trivial algorithm to the call sites. The function is internal but this is where the good stuff lives, so it better be evident how it works.

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+
+  # Pretty-print all files included in the file set.
+  # Type: (b -> String -> b) -> b -> Path -> FileSet -> b
+  _prettyFoldl' = f: start: base: tree:
```
We discussed this. This is such a clever abuse of the language that I'm not even sure if tracing should be part of the interface. It is convenient though. Maybe we could do a follow-up PR to add tracing, but first factor out the specifics of tracing from the tree fold.

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+    getInfluenceBase (union ./Makefile ../default.nix)
+    => ../.
+  */
+  getInfluenceBase = maybeFileset:
```
```suggestion
  commonPathPrefix = maybeFileset:
```

```suggestion
  commonRoot = maybeFileset:
```

I know what you mean from the descriptions, but "influence base" is not a helpful name.

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+      );
+
+  /*
+  The file containing all files from that are in any of the given file sets.
```
```suggestion
  The set containing all files that are in any of the given sets.
```

Also in the other comments I think we don't have to say "file set" every time. Such repetitiveness is quite a strain to read.

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+            predicate {
+              inherit name;
+              type = subtree;
+              ext = mapNullable head (match ".*\\.(.*)" name);
```
We'd get better orthogonality if we provided that as a library function, and an example of use in this function's documentation.

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+  A file set containing all files that are contained in a directory whose name satisfies the given predicate.
+  Only directories under the given path are checked, this is to ensure that components outside of the given path cannot influence the result.
+  Consequently this function does not accept a file set as an argument.
+  If you need to filter files in a file set based on components, use `intersect myFileSet (directoryFilter myPredicate myPath)` instead.
```
```suggestion

  If you need to filter files in a file set based on path components, use `intersect myFileSet (directoryFilter myPredicate myPath)` instead.
```

fricklerhandwerk commented on doc/functions/fileset.section.md:
```diff
@@ -0,0 +1,245 @@
+The [`lib.fileset`](#sec-functions-library-fileset) functions allow you to work with _file sets_.
+File sets efficiently represent a set of local files.
+They can easily be created and combined for complex behavior.
+Their files can also be added to the Nix store and used as a derivation source.
```
```suggestion
The contained files can also be added to the Nix store and used as a derivation source.
```

fricklerhandwerk commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+
+  - If the value is a path pointing to a directory, return a file set with all files contained in that directory
+
+  This function is mostly not needed because all functions in `lib.fileset` will implicitly apply it for arguments that are expected to be a file set.
```
```suggestion
  This function will rarely be needed because all functions in `lib.fileset` will implicitly apply it for arguments that are expected to be a file set.
```

fricklerhandwerk reviewed (approved):
The mechanics of this are really sound. Great work. Implementation guts may need a bit more clarity, especially because that's where all the magic is done it shouldn't be magical.

Most of the examples in the prose section can go to the respective functions' documentation. The derivation-related stuff absolutely needs to be documented in `mkDerivation`, otherwise it will be very hard to discover. And instead of showing how to play around with the library, the prose should briefly provide some rationale and guidance.

We could pair on the remaining text when the examples are out of the way.

infinisil committed:
Remove srcFileset stdenv.mkDerivation attribute again

As suggested by @roberth, this shouldn't be part of stdenv.mkDerivation,
it's too specialized. Also it's not that necessary, lib.fileset.toSource
and srcWorkDir are very self-explanatory

fricklerhandwerk commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -163,6 +163,31 @@ let
+    else if ! builtins.pathExists srcWorkDir then
+      throw "stdenv.mkDerivation: The `srcWorkDir` attribute \"${toString srcWorkDir}\" does not exist."
+    else if lib.filesystem.pathType srcWorkDir != "directory" then
+      throw "stdenv.mkDerivation: The `srcWorkDir` attribute \"${toString srcWorkDir}\" is not a directory"
```
Nit:

```suggestion
      throw "stdenv.mkDerivation: The `srcWorkDir` attribute \"${toString srcWorkDir}\" is not referring to a directory"
```

We use "is a" for the type of attributes, apparently.

fricklerhandwerk commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -163,6 +163,31 @@ let
+    if srcWorkDir == null then
+      null
+    else if ! builtins.isPath srcWorkDir then
+      throw "stdenv.mkDerivation: The `srcWorkDir` attribute is expected to be a path, but it's a ${builtins.typeOf srcWorkDir} instead."
```
```suggestion
      throw "stdenv.mkDerivation: The `srcWorkDir` attribute must be a path, but a ${builtins.typeOf srcWorkDir} was given."
```

A bit less verbose, and more consistent with the other errors.

fricklerhandwerk commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -163,6 +163,31 @@ let
+    else if ! src._isLibCleanSourceWith or false then
+      throw "stdenv.mkDerivation: If the `srcWorkDir` attribute is set, `src` must be a source-like value produced by the functions in `lib.sources`."
+    else if ! builtins.isPath src.origSrc then
+      throw "stdenv.mkDerivation: If the `srcWorkDir` attribute is set, `src` must be produced from a local path."
+    else if ! lib.path.hasPrefix src.origSrc srcWorkDir then
+      throw "stdenv.mkDerivation: The `srcWorkDir` \"${toString srcWorkDir}\" needs to be under the local path that produced the source \"${src.origSrc}\"."
```
```suggestion
      throw "stdenv.mkDerivation: If the `srcWorkDir` attribute is set, `src` must be produced from a local directory."
    else if ! lib.path.hasPrefix src.origSrc srcWorkDir then
      throw "stdenv.mkDerivation: The `srcWorkDir` \"${toString srcWorkDir}\" must be within the local directory that is the source of \"${src.origSrc}\"."
```

Something being "under a path" doesn't really make sense to me.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-50/29793/1

infinisil commented:
Unfortunately we received very little feedback to this idea. In order to make progress with this we have some options:
1. Open an RFC for file sets, sure to get more attention, but takes a lot of time and effort
2. Try to revive the [`lib.experimental` idea](https://github.com/NixOS/rfcs/pull/82) (with or without an RFC), then merge file set combinators as an experimental feature. I played around a bit with this idea, though it's not trivial and needs a bunch of thought.
3. Merge it (incrementally) anyways, there's no real harm to it since we're not breaking anything. This should require a number of people to review the RFC and approve of it.

Because I think this idea is really sound and useful I would like to go for option 3 so that people can use it without much more delays. Unless there's strong opposition to this, I'll now work towards this.

To start off, I will create a separate PR containing only a minimal version of the file set library with enough functionality to be useful. In particular with these functions: `coerce` (to create file sets, also implicitly), `union`/`unions` (to operate with file sets), `toSource` (to use file sets). All other functions can be added in further PRs

roberth commented:
Next time write some bugs, so people can actually make some more suggestions ;)

As the author of the original source combinators I haven't found any flaws besides not being source combinators. In fairness that's probably good though, because this only adds a new concept instead of making changes to an existing concept. So file set combinators won't interfere with existing code, period.

I'm ok with 3.

infinisil commented:
These 4 separate path-library related PR's are needed before a basic version of file set combinators can be merged, they are all up-to-date with all feedback addressed, please review:
- https://github.com/NixOS/nixpkgs/pull/243520
- https://github.com/NixOS/nixpkgs/pull/238013
- https://github.com/NixOS/nixpkgs/pull/244358
- https://github.com/NixOS/nixpkgs/pull/242695

infinisil commented:
I opened a first draft PR for just `lib.fileset.toSource` and the file set combinators base: https://github.com/NixOS/nixpkgs/pull/245623

infinisil commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+
+  # Coerce and normalise the bases of multiple file set values passed to user-facing functions
+  # Type: String -> [ { context :: String, value :: Any } ] -> { commonBase :: Path, trees :: [ <tree> ] }
+  _normaliseBase = function: list:
```
For my own future reference for once I get to implementing the combinators (in particular `intersect`):

This function here should only be used for `union`, not for `difference` or `intersect`. In particular:
- For `union`, the resulting base path should be the common prefix of all the inputs base path, e.g. `union /foo /foo/bar` should give `/foo` as the base, this is what this function does
- For `intersect`, the resulting base path should be the path with the most components of the input bases, e.g. `intersect /foo /foo/bar` should give `/foo/bar` as the base. If the paths don't overlap, an empty set is returned.
  - TODO: What base should that have? E.g. `intersect /foo /bar`, should that give an error, pick `/foo` or `/bar` arbitrarily, or use `/`? Maybe the notion of a special empty file set value should be introduced.
- For `difference, the resulting base is the same as the base from the first argument, e.g. `difference /foo /` has base `/foo`.

adrian-gierakowski commented:
Is there a lightweight way to start using these right now?  Or would I need to pull an entire copy of nixpkgs from this branch?

adrian-gierakowski commented:
> > Is there a lightweight way to start using these right now?  Or would I need to pull an entire copy of nixpkgs from this branch?
> 
> you could do a shallow copy (--depth 1) from this branch. That shouldn't be too big

thanks! So the changes from the [other PR](https://github.com/NixOS/nixpkgs/pull/222981#issuecomment-1642264905) are included here?

infinisil commented:
@adrian-gierakowski @mohe2015 I think a shallow clone would be about the same as a GitHub tarball download, I don't think there's a better way for now. In stable Nix it would look like this:
```nix
fetchTarball "https://github.com/tweag/nixpkgs/tarball/file-sets"
# Or replace `file-sets` with a commit to pin it 
```

With [niv](https://github.com/nmattia/niv) it would be
```
niv add tweag/nixpkgs --branch=file-sets
```

And with experimental Flakes:
```
{
  inputs.nixpgksFilesets.url = "github:tweag/nixpkgs/file-sets";
}
```

And if you give it a try I'd be happy to hear any feedback you have, good or bad :D

roberth commented:
You could fetch just `lib` using the tree hash:

```console
[nixpkgs]$ gh pr checkout 222981

[nixpkgs]$ git rev-parse HEAD:lib
1bdcd7fc8a6a40b2e805bad759b36e64e911036b

[nixpkgs]$ curl -L https://github.com/NixOS/nixpkgs/archive/1bdcd7fc8a6a40b2e805bad759b36e64e911036b.tar.gz | tar -tz | grep fileset
nixpkgs-1bdcd7fc8a6a40b2e805bad759b36e64e911036b/fileset.nix
```

```nix
let
  inherit (import (builtins.fetchTarball "https://github.com/NixOS/nixpkgs/archive/1bdcd7fc8a6a40b2e805bad759b36e64e911036b.tar.gz"))
    fileset;
in
  ...
```

roberth commented on lib/fileset.nix:
```diff
@@ -0,0 +1,993 @@
+
+  The predicate is called with an attribute set containing these attributes:
+
+  - `name`: The filename
```
Not 100% clear. Is this like `baseNameOf`?

infinisil replied:
Ah yup it is!

infinisil replied:
Btw note that this is the draft PR. The one ready for review is https://github.com/NixOS/nixpkgs/pull/245623

roberth commented:
> a more composable datatype

Which doesn't compose with the existing source library except through explicit conversions, and the conversion isn't isomorphic.

It's not really the composability that improves (comparatively). Rather the meaning of the operations is more set-like, to a large degree. Instead of two somewhat unintuitive attributes, we now have just one - and I'll note that in both designs we do get good behavior almost all the time.

infinisil commented:
I'll close this, because this is now mostly implemented! See https://github.com/NixOS/nixpkgs/issues/266356 for further updates and details :)
